### PR TITLE
Refactor main loop of tranquilo

### PR DESCRIPTION
**Pull request by janosg**
- [x] Redefine states such that trustregion information and model information of one state belong to same iteration
- [x] Add a success history for stagnation detection
- [x] Wrap criterion to do parallel evaluation, post processing and adding to history automatically
- [x] Add separate step for deciding how many points to sample
- [x] Add weighting to fit methods
- [x] Add weight calculation as separate step
- [x] Switch criterion of optimal sampler to fekete if we add just one point (optionally)
- [x] Add nan handling

### CODE REVIEW

File: src/estimagic/optimization/tranquilo/tranquilo.py [hunk 1]
```diff
@@ -213,78 +241,104 @@ def _tranquilo(
...
         model_fvecs = history.get_fvecs(model_indices)
 
-        centered_xs = (model_xs - trustregion.center) / trustregion.radius
+        # ==============================================================================
+        # build surrogate and optimize it
+        # ==============================================================================
+
+        weights = calculate_weights(model_xs, trustregion=state.trustregion)
+
+        centered_xs = (model_xs - state.trustregion.center) / state.trustregion.radius
```
Comment (timmens) [inline src/estimagic/optimization/tranquilo/tranquilo.py:273]: At some point, we should think about the name `centered_xs`. As we divide by the radius it is more closely related to a `standardized_xs` in my opinion. But it is not of major importance for now.

**Review by timmens — APPROVED** [review]

**Comment by timmens** [comment]
Great PR, approved. I only have a few clarifying questions.

File: src/estimagic/optimization/subsolvers/gqtpar.py [hunk 2]
```diff
@@ -70,6 +70,8 @@ def gqtpar(model, *, k_easy=0.1, k_hard=0.2, maxiter=200):
     """
     hessian_info = HessianInfo()
 
+    x_candidate = np.zeros_like(model.linear_terms)
+
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/optimization/subsolvers/gqtpar.py:74]: It did but only in rare cases while running benchmarks.

File: src/estimagic/optimization/tranquilo/filter_points.py [hunk 3]
```diff
@@ -45,6 +46,12 @@ def keep_all(xs, indices, state):
     return xs, indices
 
 
+def keep_sphere(xs, indices, state):
+    dists = np.linalg.norm(xs - state.trustregion.center, axis=1)
+    keep = dists <= state.trustregion.radius
+    return xs[keep], indices[keep]
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/optimization/tranquilo/filter_points.py:52]: Yes. Is there for trial purposes until we can specify the search norm for the scan of the history.

File: src/estimagic/optimization/tranquilo/solve_subproblem.py [hunk 4]
```diff
@@ -100,13 +100,13 @@ def get_subsolver(solver, user_options=None, bounds=None):
...
-    else:
-        bounds_dict["lower_bounds"] = getattr(bounds, "lower", None)
-        bounds_dict["upper_bounds"] = getattr(bounds, "upper", None)
+    bounds_dict = {"lower_bounds": None, "upper_bounds": None}
+    if bounds is not None:
+        for type_ in ["lower", "upper"]:
+            if hasattr(bounds, type_):
+                candidate = getattr(bounds, type_)
+                if candidate is not None and np.isfinite(candidate).any():
+                    bounds_dict[f"{type_}_bounds"] = candidate
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/optimization/tranquilo/solve_subproblem.py:109]: We don't. We just change bounds to `None` if there are no finite entries.

File: src/estimagic/optimization/tranquilo/weighting.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+from functools import partial
+
+
+def get_weighter(weighter, bounds):
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/optimization/tranquilo/weighting.py:4]: Done.

File: src/estimagic/optimization/tranquilo/tranquilo.py (hunk 1 above)
Comment (janosg) [reply src/estimagic/optimization/tranquilo/tranquilo.py:273]: Postponed.

File: src/estimagic/optimization/tranquilo/handle_infinity.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    _mins = np.min(fvecs, axis=0, where=_mask, initial=1e300)
+    _maxs = np.max(fvecs, axis=0, where=_mask, initial=-1e300)
+
+    # abs is necessary because if all values are infinite, the diffs can switch sign
+    # due to the initial value in the masked min and max
+    _diff = np.abs(_maxs - _mins)
+
+    _pos_penalty = _maxs + 2 * _diff + 1
+    _neg_penalty = _mins - 2 * _diff - 1
```
[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/optimization/tranquilo/handle_infinity.py:35]: Fixed.

**Pull request merged**